working/extension_structs/overview.md

```diff
@@ -0,0 +1,1067 @@
+  - [Protected extension types](https://github.com/dart-lang/language/issues/1467)
+
+This overview is not intended to be a feature specification: it is designed to
+give a brief overview of the core proposal and set out some design points for
+discussion.  If there is buy-in on these ideas, we may choose to incorporate
+them into an existing proposal, or add a new feature specification.
+
+
+## Design principles
+
+We aim to minimize the differences between structs and classes.  As much as
+possible, structs should behave as restrictions of classes, and extension
+structs should behave as further restrictions of structs.  We specifically aim
+to avoid as much as possible having different behaviors for the same concept
+(e.g. differences in scoping).
+
+We also aim to minimize the amount of new syntax required, and to maximize the
+amount of new functionality that we can provide relative to the syntactic real
+estate consumed, and the new cognitive load imposed on users.
+
+Concretely for the purposes of this proposal, we have started with the following
+goals:
+  - If a struct or an extension struct `B` says that it `implements` or
+    `extends` `A`, then:
+    - It should be the case that `B` is a subtype of `A`
+    - It should be the case that `B` has a superset of the method
+      names/signatures of `A`
+    - It should be the case that assigning an instance of `B` to a location of
+      type `A`, while it may change the set of members available, should not
+      change the dispatch of any members (that is, if `f` is available on `B`,
+      then calling `f` through either interface reaches the same code).
+
+
+It is not clear that all of these goals can be met while meeting requirements.
+In particular, the last goal is incompatible with overriding methods, given our
+intended semantics, and it is likely that allowing some form of overriding is a
+requirement.
+
+## Roadmap
+
+The first section describes the full struct feature.  The second section
+describes the restriction of the full struct to support static wrappers.
+
+## Structs
+
+This section describes how structs look and behave, largely by example.  In many
+cases, alternatives or extensions to the core proposal are described in line.
+Larger extensions are postponed to a later section.
+
+## Structs by example
+
+Structs allow you to define simple classes containing immutable data very
+compactly.  For example, we might model a component vector as follows:
+
+```
+struct Component(int x, int y, int z, int w = 1);
+```
+
+This short definition essentially defines a class `Component`, with four fields,
+and provides a number of convenient methods.
+
+```dart
+void test() {
+  // Call the default constructor, using the default value for w
+  var c1 = Component(0, 0, 0);
+  // Call the default constructor, explicitly passing w
+  var c2 = Component(1, 1, 1, 1);
+
+  assert(c1 == c2); // Equality is defined to be structural
```

Comment:
Should this be `Component(0, 0, 0, 1)` for a structural equality to match `Component(0, 0, 0)`?